### PR TITLE
HA mailbox, distributed lock

### PR DESCRIPTION
**Pull request by spoon16**
To enable a highly available (or horizontally scalable) mailbox we have eliminated as much local state management as possible. Coordination and the synchronization of mailbox events via locking represents one of the most pervasive and complicated parts of legacy mailbox state management. This PR addresses the distribution of local coordination via thread locking by implementing distributed coordination via an external coordinator (Redis).

Legacy mailbox worker coordination happened locally to each mailbox process. That represents a problem when you move to a mailbox worker pool model. Mailbox requests will be routed to any available mailbox worker, each mailbox worker lacks the ability to coordinate the application of side affects. Many race conditions that are difficult to debug can manifest. For instance a customer "sends" a message and then immediately clicks on their "sent items" folder to double check what they wrote. Without coordination the request to load the contents of the "sent items" folder may be processed by a server that has no record of the recently sent message, resulting in a confusing user experience.

F9 Teams has refactored `com.zimbra.common.mailbox.MailboxLock` converting it from concrete implementation to interface. `MailboxLock implements AutoClosable` so that it can be used conveniently with Java's [try-with-resources statement](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html). We have introduced `com.zimbra.common.mailbox.MailboxLockFactory` which allows lock references to be dynamic rather than static. The concrete implementation and logic that used to be in `MailboxLock` has been moved to `com.zimbra.cs.mailbox.LocalMailboxLockFactory` and `com.zimbra.cs.mailbox.LocalMailboxLockFactory.LocalMailboxLock`.

A distributed lock that leverages Redis via a dependency on [Reddison](https://github.com/redisson/redisson) has been implemented `com.zimbra.cs.mailbox.DistributedMailboxLockFactory` and `com.zimbra.cs.mailbox.DistributedMailboxLockFactory.DistributedMailboxLock`. The `DistributedMailboxLockFactory` and `DistributedMailboxLock` pass all relevant unit tests and SOAP Sanity test suite tests. It is a 🐛 if `DistributedMailboxLock` behaves differently than `LocalMailboxLock`.

The approach we have taken appears to synchronize events across a mailbox worker pool in a very conservative way, it clones the legacy behavior. It is very likely that in a HA configuration request performance will be substantially worse than in a legacy configuration because of overuse of distributed locking. Our principle is illustrated well by this quote from Google's Spanner team, "we believe it is better to have application programmers deal with performance problems due to overuse of transactions as bottlenecks arise, rather than always coding around the lack of transactions". Profiling this implementation will be straightforward and there are many opportunities for optimization.

This diagram demonstrates how requests are coordinated across a mailbox worker pool with two mailbox workers. When a request comes into a mailbox worker and another request is already being processed for the same mailbox by another mailbox worker processing will be delayed until the other mailbox worker releases the distributed lock.

<p align="center">
  distributed lock timeline
</p>

`com.zimbra.cs.mailbox.Mailbox#beginTransaction` and `com.zimbra.cs.mailbox.Mailbox#endTransaction` were also impacted by this work. These methods have been refactored into `com.zimbra.cs.mailbox.Mailbox.MailboxTransaction` an abstraction that encapsulates and manages the life cycle of a mailbox transaction. `MailboxTransaction implements AutoClosable`  so that it can be used conveniently with Java's try-with-resources statement. Instead of passing `boolean success` to `endTransaction` a call to `MailboxTransaction#commit` or `MailboxTransaction#rollback` should be made before `MailboxTransaction#close`, rolling back is the default behavior.

### CODE REVIEW

File: client/src/java/com/zimbra/client/ZLocalMailboxLockFactory.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    @Override
+    public MailboxLock lock(boolean write) {
+        if (write) {
+            return writeLock();
+        }
+        return this.readLock();
+    }
+
+    @Override
+    public void close() throws Exception {
```
[the remark this answers is not in the document]
Comment (Raffaell0) [reply client/src/java/com/zimbra/client/ZLocalMailboxLockFactory.java]: No always is required a implementation in Factory level for close() method, this case is not applicable

File: store/src/java-test/com/zimbra/cs/mailbox/DistributedMailboxLockTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	        final long sleepTime = 10;
+	        int joinTimeout = 10000;
+
+	        List<Thread> threads = new ArrayList<Thread>(numThreads * 2);
+	        for (int i = 0; i < numThreads; i++) {
+	            String threadName = "MailboxLockTest-MultiReader-" + i;
+	            Thread reader = new Thread(threadName) {
+	                @Override
+	                public void run() {
+	                    for (int i = 0; i < loopCount; i++) {
```
[the remark this answers is not in the document]
Comment (Raffaell0) [reply store/src/java-test/com/zimbra/cs/mailbox/DistributedMailboxLockTest.java]: This unit test is based on MailboxLockTest.multiAccess counter is the same logic, I think we can delete this complete file

File: store/src/java/com/zimbra/cs/mailbox/MailboxManager.java [hunk 3]
```diff
@@ -860,11 +863,12 @@ private synchronized Mailbox createMailboxInternal(OperationContext octxt, Accou
         CreateMailbox redoRecorder = new CreateMailbox(account.getId());
 
         Mailbox mbox = null;
-        boolean success = false;
+        Mailbox.MailboxTransaction mboxTransaction = null;
```
[the remark this answers is not in the document]
Comment (Raffaell0) [reply store/src/java/com/zimbra/cs/mailbox/MailboxManager.java]: Doesn't apply, MailboxTransaction is not a static class

File: store/src/java/com/zimbra/cs/mailbox/LocalMailboxLockFactory.java [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+            return zLock.hasQueuedThreads();
+        }
+
+        /**
+         * Acquires the lock.
+         *
+         * @throws LockFailedException failed to lock
+         */
+        @Override
+        public void lock() {
```
[the remark this answers is not in the document]
Comment (Raffaell0) [reply store/src/java/com/zimbra/cs/mailbox/LocalMailboxLockFactory.java]: There are 5 conditions to check, including asserts calls, assert calls use 2 conditions to check and waiters one more, these two last were added by unit tests, in a "productive" code, the better case should be to check one condition and in the worst scenario 3, I think is acceptable

File: soap/src/java/com/zimbra/soap/admin/message/GetDelegatedAdminConstraintsResponse.java [hunk 5]
```diff
@@ -53,6 +54,7 @@ public void setAttrs(Iterable <ConstraintAttr> attrs) {
 
     public GetDelegatedAdminConstraintsResponse addAttr(ConstraintAttr attr) {
         this.attrs.add(attr);
+        ZimbraLog.mailbox.info("----------- >>>>>>>>>"+  attr.toString());
```
Comment (spoon16) [inline soap/src/java/com/zimbra/soap/admin/message/GetDelegatedAdminConstraintsResponse.java]: this should be removed

File: store/src/java/com/zimbra/cs/session/Session.java [hunk 6]
```diff
@@ -172,7 +173,8 @@ public Session unregister() {
         if (null != mboxStore) {
             if (mboxStore instanceof Mailbox) {
                 Mailbox mbox = (Mailbox)mboxStore;
-                assert(mbox.lock.isWriteLockedByCurrentThread() || !Thread.holdsLock(this));
+             // @Raffaell0 not all callers will be holding a mailbox lock, no reasonable way to pipe in MailboxLock instances here for this assertion
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/session/Session.java]: remove this comment

**Comment by gordyt** [comment]
Closing this pull request as I will be doing some maintenance on the branches in `zm-mailbox`. In particular:

1. I am renaming this pull request branch from `feature/ha` to `feature/ha-distributed-lock`
2. I am renaming the `feature/ha-combined` branch to `feature/ha` so that it will be consistent with the other HA-related branches that we are building off of.

**Pull request closed without merging**